Admincli edits

leswilliams44 opened the pull request:
Completed the first pass on Chapter 19, Admin-CLI. Main changes include:
* changing tense to present
* writing in the second person, imperative mood
* adding non-visible AsciiDoc syntax
* following IBM Style Guide conventions
* adding articles and nouns where necessary for clarification

leswilliams44 committed:
first pass at editing Sections 19.1 through 19.6

leswilliams44 committed:
first complete pass at editing Chapter 19, admin-cli

matthewhelmke commented:
Thank you, @leswilliams44 . I appreciate your attention to detail and kind assistance. Looks good to me.

matthewhelmke commented:
@stianst This is ready to merge. I just wanted to make sure you had a chance to have a dev team member review, if desired. The changes were all linguistic and should not impact techinical content.

stianst commented:
@mstruk do you want to check these edits?

mstruk commented:
@stianst Yes, I'll check today.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+
+The Admin CLI works by making HTTP requests to Admin REST endpoints. Access to them is protected and requires authentication.
+
+//Should there be a link for the REST API documentation?
```
It would be nice to have a link. The reason I have not included one is most likely that I wasn't sure we have a canonical url that won't change in the future. By avoiding the link we avoid a possibility of having a broken link.

matthewhelmke replied:
Sounds reasonable. Is it standard practice in this project to leave the comment in the file and replace it in the future when/if we have a canonical URL for the REST API documentation, or does it need to be removed?

mstruk replied:
It's not standard practice but I did it once or twice as well. It seems leaving it there doesn't actually increase the probability of acting to address it :)
We have a build plugin that checks external links, and it's failing all the time therefore we already have (maybe many) broken links in our docs.

Maybe best if @stianst chimes in on this one.

leswilliams44 replied:
I will remove the link.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+c:\> kcadm get clients/%CID%/installation/providers/keycloak-oidc-keycloak-json
+----
+
+. In a production environment, you access {project_name} with `https:` to avoid exposing tokens to network sniffers. If a server's certificate is not issued by one of the trusted certificate authorities (CAs) that are included in Java's default certificate truststore, prepare a [filename]`truststore.jks` file and instruct the Admin CLI to use it.
```
I would change 'you access' to 'you have to access'.

matthewhelmke replied:
Linguistically, this is a more nuanced question. On the one hand, the phrase "...you access..." is a subject-verb phrase that is direct and clear, which is a hallmark of good technical documentation. On the other hand, the content is actually intending to communicate that "https:" must be used. While "you access" is correct, "you must access" is clearer.

"you have to access" is correct in terms of meaning, but "you must access" is a stronger standard in technical documentation, so I suggest using "you must access."

mstruk replied:
They both sound good to me.

leswilliams44 replied:
Revised to use "you must access"

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
-or alternatively you can specify only a client id, which will result in special, so called 'service account' being used. In the first case,
-a password for the specified user has to be used at login. In the latter case there is no user password - only client secret.
-Alternatively, `Signed JWT` can be used.
+When you log in with the Admin CLI, you specify a server endpoint URL and a realm, and then you specify a username. Another option is to specify only a client ID, which results in using a special "service account". When you log in using a username, you must use a password for the specified user. When you log in using a client ID, you only need the client secret, not the user password. You could a ...
```
Same here as with Client Registration CLI, it would be better to use 'client id' instead of 'client ID' as it's a colloquial reference to 'clientId' attribute, rather than 'id' attribute.

matthewhelmke replied:
The style guide used for our documentation requires "ID" to always be capitalized, both letters. So, if we are talking about a generic "client ID," what I have written in this sentence is what we should use.

If we are talking about a specific attribute or variable, we should spell it exactly as written/required in the command, 'clientId'. We are not doing that here.

While reading this, I think we can make the section clearer. First, are these two login methods used frequently, say each about 50% of the time? Or, is one significantly more common?

If they are used with about the same frequency, we should introduce them with a phrase and put each in a bullet, like this:

There are two ways to log in using the Admin CLI:
* With a username and password - details...
* With a client ID - details...

If one is used more frequently and the other is more of a side case, I would describe the most-used option and then put the secondary option in a NOTE box.

tanberry replied:
Yay, a robust conversation about grammar and nouns and words! I want to join in. ;-) So on this one, I have to say that because we are not actually talking about the attribute (clientId), then we do have to capitalize 'ID'. We are talking about a noun that is a value, but in this sentence we are not talking about the attribute. We *could* of course change the sentence to talk about the attribute, and then use "clientId" (i.e. "When you log in using the clientId attribute, you only need the...").

I also suggest that we separate out the sentences about client ID, because it is smack in the middle of sentences about user name. I suggest reworking the whole paragraph to either provide the two options (username vs. client ID) as bullets, or if using client ID is the least common use case, then focus on the user name and add a Note below about the client ID. But they are mixed up too much in current paragraph, for sure.

@mstruk do we know whether most users use username or client ID?

And, sorry to say, unless you are talking about the attribute called 'username', it needs to be two separate words. ;-)

mstruk replied:
@tanberry @matthewhelmke The issue here is that the text in fact is not talking about a client ID as far as I understand what ID means. Client has two distinct attributes that can both be used as an ID. One is called 'id' which is a 30 chars long random string (UUID actually) and is not very practical to use. The other is called 'clientId' which is a human readable name - but a logical name, not a caption or descriptive name for display. This name does not change so it can be used as ID. We use the second one as ID in this case. Thus talking about ID rather than clientId obfuscates things. While in this particular case we don't use 'id' attribute, it is in fact used in other cases where we also refer to it as ID. Thus when we say ID we sometimes refer to 'id', and sometimes to 'clientId' attribute. That is the point I was trying to highlight here. There would be no misunderstanding if we could just use 'clientId' when we refer to 'clientId' attribute. For readability I used 'client id' in several places instead of 'clientId'. But that was apparently misleading already as it resulted in being replaced as client ID which obfuscates it even more.

matthewhelmke replied:
I see. I didn't understand that until now. Perhaps we should just call this 'clientId' everywhere. What do you think?

leswilliams44 replied:
I will use clientId unless told otherwise.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
 
-This approach maintains an authenticated session between `kcadm` command invocations by saving the obtained access token, and
-associated refresh token, possibly other secrets as well in a private configuration file. See <<_working_with_alternative_configurations, next chapter>> for more info on configuration file.
+For example, when performing an command, we specify all the information required for authentication.
```
type: 'an command' -> 'a command'

matthewhelmke replied:
The original sentence was on line 92 and read "For example, when performing an operation we specify all the information required for authentication:".

I am curious whether changing "operation" to "command" changes the meaning more than would be preferred. I think it might. Rather than fix the "an" to an "a" perhaps we change back to "operation."

We issue a command to perform an operation...

tanberry replied:
@leswilliams44 where did this new sentence with "an command" come from? Was it copied/pasted from elsewhere?

mstruk replied:
To me 'command' and 'operation' mean the same. Go with whatever sounds better.

leswilliams44 replied:
"an command" was changed back to "an operation"

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
 
-By default, `Admin CLI` automatically maintains a configuration file called `kcadm.config` located under user's home
-directory - it's full pathname is `$HOME/.keycloak/kcadm.config` (on Windows it's `%HOMEPATH%\.keycloak\kcadm.config`).
+By default, the Admin CLI automatically maintains a configuration file called [filename]`kcadm.config` located under the user's home directory, which is the full path name [filename]`$HOME/.keycloak/kcadm.config`. On Windows, the full path name is [filename]`%HOMEPATH%\.keycloak\kcadm.config`.
```
I think instead of 'which is the full path name ...' it should be 'with the full path name ...'

matthewhelmke replied:
First off, we are not listing the full path name. There is a variable in there that stands in for /home/username. The true full path name would be /home/username/.keycloak/kcadm.config.

I suggest rephrasing to say this:

By default, the Admin CLI automatically maintains a configuration file called [filename]`kcadm.config` located at [filename]`$HOME/.keycloak/kcadm.config`. On Windows, the file is located at [filename]`%HOMEPATH%\.keycloak\kcadm.config`.

UNLESS there is a reason that the true, full path name is actually required rather than the version with the variable...in that case, we need to edit to list the full path name.

tanberry replied:
I would separate this sentence into two sentences: "By default, the Admin CLI automatically maintains a configuration file called [filename]`kcadm.config` located under the user's home directory. In Linux-based systems, the full path name is [filename]`$HOME/.keycloak/kcadm.config`. On Windows, the full path name is [filename]`%HOMEPATH%\.keycloak\kcadm.config`.

However, can we really say "full" path name?? Do we even know the full path name, always?

It looks like [filename] is a variable, not sure how that will work in front of the full path name? 

Also, the phrase "automatically maintains a config file" is unclear to me. Does "maintain" mean that the config file is automatically written to and also continually updated? Or does it only meant that the config file is created automatically, but not written to? Can it be manually edited? If it is important to call out that the file is maintained (written to), and if it is important to note that the file can not be manually edited, then let's say that.

mstruk replied:
It doesn't have to say full path name. 'Located at' sounds good to me.

@tanberry Yes, 'maintained' means it is written to, and updated. As session progresses, the access token reaches it's time to live, and is automatically refreshed. The new one is saved in the file. It can be manually edited because it's a JSON file but I don't see a use case where that would be needed. We don't describe file's format, and we don't instruct users to change it directly. The important thing here is to point out that this file exists, and that it contains some secrets and so it should not be readable to other users.

leswilliams44 replied:
I used the two sentences suggested by @tanberry.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
-Make sure to not make config file visible to other users on the system as it contains access tokens, and secrets that should be kept private.
-By default the ~/.keycloak directory and its content will be automatically created with proper access limits. However if directory will
-exist already with non-default permissions, those will not be updated.
+Make sure you do not make the configuration file visible to other users on the system. It contains access tokens and secrets that should be kept private. By default, the [filename]`~/.keycloak` directory and its content are created automatically with proper access limits. If the directory already exists with non-default permissions, those access tokens and secrets will not be updated.
```
The last sentence is semantically incorrect. If the directory already exists then permissions won't be updated (rather than tokens and secrets).
I suggest: 'If the directory already exists, its permissions won't be modified.'

matthewhelmke replied:
There is a little more information required. The "with non-default permissions" is a pretty important bit of info. 😄 How about this?

If the directory already exists and its permissions are different from the {project_name} default permissions, those permissions will not be reset. In this case, it is recommended that you check and change the permissions to the {project_name} defaults.

tanberry replied:
As good tech writers, we avoid the future tense and instead write in present tense. So I suggest "If the directory already exists with non-default permissions, those permission are not updated."

mstruk replied:
@matthewhelmke Default or non-default permission on the directory is not a factor. Mere existence of the directory means its permissions won't be touched. Non-default permissions is actually misleading the way I used it and should be removed.

We're not talking about Keycloak server here so this is not the right context for using {project_name}.

@tanberry What about simply 'If the directory already exists, its permissions are not updated.'

leswilliams44 replied:
Used "If the directory already exists, its permissions are not updated.'"

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
-You may want to avoid storing any secrets at all inside a config file for the price of less convenience and having to do more token requests.
-In that case you can use `--no-config` option with all your commands. In that case you will have to specify all the authentication info needed by
-`config credentials` command with each `kcadm` invocation.
+Consider avoiding storing secrets inside a configuration file, even though it is less convenient and you have to make more token requests. If you choose to not store secrets, you can use the [command]`--no-config` option with all your commands and specify all the authentication information needed by the [command]`config credentials` command with each [command]`kcadm` invocation.
```
Using 'Consider' here sounds as a suggestion - a hint that it's a better option. I don't think it is a better option. This support is only here for the extra paranoid, and it will probably end up being used by only a very tiny percentage of users. In fact another use-case is when you only need to perform a single operation, and it's more efficient to do it in one single step.

matthewhelmke replied:
That is good information! Does this express it more clearly?

If your unique circumstances require you to avoid storing secrets inside a configuration file, you can do so. It will be less convenient and you will have to make more token requests. To not store secrets, use the [command]`--no-config` option with all your commands and specify all the authentication information needed by the [command]`config credentials` command with each [command]`kcadm` invocation.

tanberry replied:
Agree with @mstruk that the current phrasing makes it sound like a good idea. I would rearrange the words (and maybe make is a Note) to say:

 "If you choose to not store secrets inside a configuration file, you can use the [command]`--no-config` option with all your commands and specify all of the authentication information that is needed by the [command]`config credentials` command with each [command]`kcadm` invocation. Be aware that this approach is less convenient and you have to make more token requests.  

However,  the [command]`--no-config` option is useful for when a only single operation is needed."

mstruk replied:
I like @matthewhelmke suggestion better (sorry @tanberry :)

leswilliams44 replied:
Used @matthewhelmke's suggestion

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+----
 
-Main usage pattern is the following:
+ENDPOINT is a target resource URI and can either be absolute (starting with `http:` or `https:`) or relative, used to compose an absolute URL of the following format.
```
I think this sentence should end with a colon - ':'

matthewhelmke replied:
👍

tanberry replied:
agree...

leswilliams44 replied:
added a colon

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
-Realm is not enabled by default. By enabling it, it can be used for authentication immediately.
-
-A description for a new object can be in JSON format as well:
+==== Creating a new realm
```
My concern around replacing 'Subtitle::' notation with '==== Subtitle' notation is that the second one results in all these subtitles being displayed in navigation in the left panel. The number of these titles is significant, and they will dominate the whole document's navigation panel, making it also aesthetically displeasing. I was unable to find a way to suppress the fourth level of depth in navigation, and that is the reason 'Subtitle::' notation was used (using labels instead of subtitles).

matthewhelmke replied:
It is always tough to maintain that delicate balance between aesthetics and usefulness. The hard question that I can't yet answer for this project is whether our users will know to look in the '===Realm operations' section to find the info they need or whether they need the sign posts provided by a heading listed in the sidebar.

In the short term, I will defer to @mstruk's opinion here, since he is much more familiar with this project and I'm the newb. :)

mstruk replied:
Don't get me wrong - I know that using labels instead of subtitles is a hack. The real fix is to have a config setting that allows me to control to what depth subtitles should be included in navigation, and allow such control on a per-chapter basis. But I think we first need to find this fix before we can merge this change of labels to subtitles.

matthewhelmke replied:
👍

leswilliams44 replied:
@mstruk Is this a possible solution?  https://stackoverflow.com/questions/34481638/how-to-use-tocify-with-asciidoctor-for-a-dynamic-toc

leswilliams44 replied:
Here is another option.

If the title text should not be numbered or appear in the table of contents, use a heading at the appropriate level, and then set the discrete attribute above the heading.

[discrete]
==== Delete a realm
Text text text...

mstruk replied:
Fantastic, that's it!
I tried it and it works. Do you want to add that to all fourth level subtitles in admin-cli.adoc?

matthewhelmke replied:
👍

matthewhelmke replied:
@stianst , who do we need to talk to to implement this solution that @leswilliams44 found above and @mstruk has tested? The solution is being suggested because this is a very long page that has too many entries in the table of contents. The ultimate solution would be for me to break this up into multiple pages, but that isn't something I have bandwidth for today (but it is on my TODO list for the future).

https://stackoverflow.com/questions/34481638/how-to-use-tocify-with-asciidoctor-for-a-dynamic-toc

Essentially, it involves adding a js file (full source in the link above) and then running asciidoctor with different options to build.

So, it looks like a great solution for asciidoctor, but I don't know how or even if this is feasible with our current toolchain. If so, I would like us to consider it. If not, I wonder if there is another option.

Do you know, and if not, do you know who I should be asking? Thanks.

mstruk replied:
@matthewhelmke The [discrete] solution doesn't require anything extra in the build. It is supported out of the box. All that has to be done is to insert [discrete]  in front of all fourth level subtitles in admin-cli.adoc.

matthewhelmke replied:
@mstruk OH! I totally misread the earlier comment. Yes! Let's just use [discrete]. That is elegant and simple. Thank you for correcting my understanding.

leswilliams44 replied:
I will apply the [discrete] attribute to all level four headings.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Getting a specific realm
 
-    $ kcadm.sh get realms --fields realm --format csv --noquotes
+You append an ID to a collection URI to get an individual item of a collection, as is common for REST web services.
```
Actually this sentence is dubious, and was dubious in the first place. It's an aberration that for realm we don't use 'id' field, but 'realm' field. Saying we add a realm id is thus misleading. 
I suggest: 'You append a realm name to a collection URI to get an individual realm.'

matthewhelmke replied:
👍

leswilliams44 replied:
Revised to say "You append a realm name to a collection URI to get an individual realm."

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+----
+$ kcadm.sh get clients -r demorealm --fields id,clientId
+----
+. Create a new role by using the client's ID attribute to construct an endpoint URI, such as [filename]`clients/ID/roles`.
```
Throughout this document I see consistent replacement of 'id' with ID, but at this point here it really makes it painfully obvious to me that it's just wrong. JSON is case sensitive, and has no notion of primary key. Saying 'client's ID attribute' is just wrong. It should really be 'client's 'id' attribute'.

What is the rationale for this consistent replacement of 'id' with ID?
I see in the rest of the document, that where formulation is "'id' attribute" it was not changed into ID. In fact it looks like all occurences of ID can be replaced with "'id' attribute". That would make everything more clear and consistent.

matthewhelmke replied:
If it is in code, everything needs to be exact.

If it is in English, then the standard in our adopted style guide is that "ID" is always capitalized.

![2017-12-11 14 37 51](https://user-images.githubusercontent.com/660205/33852680-123ae124-de81-11e7-8d80-e555c38b4bc4.jpg)

mstruk replied:
What about English that refers to code which is how we use ID throughout this document?

You can't refer to 'clientId' attribute in JSON document by calling it client ID. But that's what 'client id' as used in many places actually refers to - to 'clientId' attribute in a JSON document. Imagine that we're talking about a product record structured with fields: SKU, Name, Price. Then throughout the document you refer to SKU by constantly saying ID.
Does that make any sense?

matthewhelmke replied:
I agree. I never want to refer to a SKU by saying ID. If that is what we are doing, we need to fix it.

Any time we refer specifically to code, we absolutely must use the exact same string as is in the code, such as 'clientId'. When we are speaking generically, we should use 'ID'. For example, we might say this: Refer to 'clientId' to determine the client ID. Okay, that is a weak sentence, but I hope it adequately illustrates the difference.

leswilliams44 replied:
I will go through the entire document and correct the revisions to clientID to the best of my understanding. Please check my work.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
-Getting a specific realm role::
-
-Use `get` command, and role `name` to construct an endpoint uri for a specific realm role - `roles/ROLE_NAME`
+Use the [command]`get` command and the role [filename]`name` to construct an endpoint UIR for a specific realm role: [filename]`roles/ROLE_NAME`, where [filename]`user` is the name of the existing role.
```
Typo: 'endpoint UIR' -> 'endpoint URI'

matthewhelmke replied:
👍

leswilliams44 replied:
Corrected typo

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+----
+$ kcadm.sh get-roles -r demorealm --cclientid test-client --rolename operations
+----
+. Assume that there is a client with client ID of [filename]`test-client`, a client role called [filename]`support`, and another client role, which becomes a composite role, that has an ID of `"fc400897-ef6a-4e8c-872b-1581b7fa8a71", "name":"operations"`.
```
There's something wrong with the ending of this sentence. 

I suggest:
'..., and another client role called [filename]`operations`, which becomes a composite role, that has an ID of  "fc400897-ef6a-4e8c-872b-1581b7fa8a71"'.

matthewhelmke replied:
That is clearer. Thank you for helping sort out the technical bits!

leswilliams44 replied:
The suggested change has been made.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+----
+. Assume that there is a client with client ID of [filename]`test-client`, a client role called [filename]`support`, and another client role, which becomes a composite role, that has an ID of `"fc400897-ef6a-4e8c-872b-1581b7fa8a71", "name":"operations"`.
++
+In this example, [filename]`operations` is the target composite role, and the role's ID precedes [filename]`operations`.
```
I don't understand the 'precedes' part of this sentence nor what example this sentence refers to. Also seems like it doesn't fit right with the next sentence.

I think this sentence can safely be removed, as all the relevant information is given in the previous sentence.

matthewhelmke replied:
Is this form used anywhere?
'"name":"operations"'
If not, I agree. If so, is it introduced elsewhere/is it important?

mstruk replied:
No, this "name":"operations" was an error in the first place and is fixed with the previous suggestion.

matthewhelmke replied:
👍

leswilliams44 replied:
The sentence has been deleted.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Adding client roles to a realm role
 
-In this example 'operations' is our target composite role, and we just got its `id`.
+You can create or modify a composite realm role.
```
I think reworked text makes this sentence redundant. Originally it is simply an intro into a recipe how to do that. When it stands alone like this it seems to suggest that you can create or modify, but maybe not delete and read, which is not true. I suggest removing it.

matthewhelmke replied:
👍

leswilliams44 replied:
The sentence has been deleted.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Adding client roles to a client role
 
-    $ kcadm.sh get-roles --rid fc400897-ef6a-4e8c-872b-1581b7fa8a71 --all
+You can create or modify a composite client role.
```
Again, this sentence is redundant.

matthewhelmke replied:
👍

leswilliams44 replied:
The sentence has been deleted.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
Second sentence is incorrect. It is not admin who will change the password and who is the intended audience of this document. Rather it is one of the users who will have to change their password when they log in the next time.

I suggest: 'The target user will have to change the password the next time they log in.'

matthewhelmke replied:
👍

leswilliams44 replied:
The suggested change has been implemented.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
This sentence doesn't sound quite right. Maybe a slight rewording:

'You can use the [command]`--userid` to specify the user by the value of [command]`id` attribute.'

matthewhelmke replied:
The original said, "You can use `--userid` if you want to specify the user by using `id` attribute." This was clear and accurate and I think we can go back to the original phrasing.

leswilliams44 replied:
@matthewhelmke's suggestion was implemented.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
Last sentence should rather be: "The target user will have to change the password the next time they log in."

matthewhelmke replied:
👍

leswilliams44 replied:
The change has been made.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
I think it sounds better to say: "... on the [command]`realm management` client ...".

matthewhelmke replied:
I agree. (For convenience, this is to replace "For example, to remove from user `testuser` two roles defined on client `realm management`...", line 1069.)

leswilliams44 replied:
The change has been implemented.

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
I think it sounds better to say: "... on the [command]realm management client ...".

matthewhelmke replied:
Line 1080. I agree.

leswilliams44 replied:
done

mstruk commented on server_admin/topics/admin-cli.adoc:
```diff
@@ -1,1320 +1,1702 @@
+==== Listing users
 
-Use `update` operation with the same endpoint uri as for getting a specific user. For example on Linux:
+Use the [filename]`users
```
Typo - repeated word: 'If you want want'

matthewhelmke replied:
Line 1665. Nice catch!

leswilliams44 replied:
typo corrected

mstruk commented:
For some reason line numbers on my suggestions are invalid. That sucks big time.

matthewhelmke commented:
@mstruk I went through and found the line numbers that are missing in the later comments. :)

tanberry reviewed (commented):
It's great to see all the edits and the edits of the edits, so I added some edits of the edits' edits. ;-) Great stuff.

leswilliams44 commented:
Stopped at line 525. Will continue to address all comments.

leswilliams44 committed:
ddressing comments from the first editorial review.

leswilliams44 committed:
Finished editing revisions for Admin CLI

I went through all of the recommendations and implemented them. Please let me know if I did not interpret the recommendations correctly.
